source/non_matching/mesh_classifier.cc

```diff
@@ -0,0 +1,531 @@
+                                    const FiniteElement<dim> &element);
+
+        /**
+         * Returns the finite element passed to the constructor wrapped in a
+         * collection.
+         */
+        const hp::FECollection<dim> &
+        get_fe_collection() const override;
+
+        /**
+         * Returns 0, since there is always a single element in the
+         * FECollection.
+         */
+        unsigned int
+        active_fe_index(const typename Triangulation<dim>::active_cell_iterator
+                          &cell) const override;
+
+        /**
+         * Return the level set function evaluated at the real space support
+         * points of the finite element passed to the constructor.
+         */
+        void
+        get_local_level_set_values(
+          const typename Triangulation<dim>::active_cell_iterator &cell,
+          Vector<double> &local_levelset_values) override;
+
+        /**
+         * Return the level set function evaluated at the real space face
+         * support points of the finite element passed to the constructor.
+         */
+        void
+        get_local_level_set_values(
+          const typename Triangulation<dim>::active_cell_iterator &cell,
+          const unsigned int                                       face_index,
+          Vector<double> &local_levelset_values) override;
+
+      private:
+        /**
+         * Pointer to the level set function.
+         */
+        const SmartPointer<const Function<dim>> level_set;
+
+        /**
+         * Collection containing the single element which we locally interpolate
+         * the level set function onto.
+         */
+        const hp::FECollection<dim> fe_collection;
+
+        /**
+         * FEValues object used to transform the support points on a cell to
+         * real space.
+         */
+        FEValues<dim> fe_values;
+
+        /**
+         * FEFaceValues object used to transform the support points on a face to
+         * real space.
+         */
+        FEFaceValues<dim> fe_face_values;
+      };
+
+
+
+      template <int dim>
+      AnalyticLevelSetDescription<dim>::AnalyticLevelSetDescription(
+        const Function<dim> &     level_set,
+        const FiniteElement<dim> &element)
+        : level_set(&level_set)
+        , fe_collection(element)
+        , fe_values(element,
+                    Quadrature<dim>(element.get_unit_support_points()),
+                    update_quadrature_points)
+        , fe_face_values(element,
+                         Quadrature<dim - 1>(
+                           element.get_unit_face_support_points()),
+                         update_quadrature_points)
+      {}
+
+
+
+      template <int dim>
+      void
+      AnalyticLevelSetDescription<dim>::get_local_level_set_values(
+        const typename Triangulation<dim>::active_cell_iterator &cell,
+        Vector<double> &local_levelset_values)
+      {
+        fe_values.reinit(cell);
```

Comment:
Can you please add an assertion,
```suggestion
        AssertDimension(local_levelset_values.size(), fe_values.n_quadrature_points);
        fe_values.reinit(cell);
```

Reply:
This function was removed, but I added a similar assert in the `get_local_level_set_values` for faces.